extensions/interactions/PencilCodeEditor/directives/oppia-interactive-pencil-code-editor.component.ts

```diff
@@ -158,50 +172,66 @@ export class PencilCodeEditor implements OnInit, OnDestroy {
+    // It is passed to both `execute` and `registerCurrentInteraction()`
+    // to ensure that submission logic remains consistent.
 
-    this.pce.on('execute', () => {
+    let submitInteractionAnswer = () => {
+      // Prevents submission if an error has occurred or an answer has already been submitted.
       if (errorIsHappening || hasSubmittedAnswer) {
         return;
       }
-      // The first argument in the method below gets executed in the
-      // pencilcode output-frame iframe context. The code input by the
-      // user is sanitized by pencilcode so there is no security
-      // issue in this case.
+
+      // Evaluates the code inside the Pencil Code iframe's context.
+      // The first argument 'document.body.innerHTML' retrieves the current
+      // HTML output generated by the user's code execution.
+      // PencilCode sanitizes user input, ensuring security.
       this.pce.eval(
         'document.body.innerHTML', // disable-bad-pattern-check
         (pencilCodeHtml: string) => {
+          // Normalize the user's inputted code (e.g., replace tabs with spaces).
           let normalizedCode = this.getNormalizedCode();
 
-          // Get all the divs, and extract their textual content.
+          // Extracts all div elements from the executed output to get the textual content.
           let temp = document.createElement('div');
-          // 'pencilCodeHtml' here is a string of raw code for div of
-          // 'pencil-code-editor-iframe' in template. In order to extract all
-          // divs from raw code we are converting it first into an element
-          // thereby selecting all the divs inside it and then extracting all
-          // the content inside them for which 'innerHTML' is used.
+
+          // `pencilCodeHtml` is a string representing the raw HTML content inside
+          // the output frame. We convert it into an element to access and extract
+          // textual content from all the `div` elements present.
           // eslint-disable-next-line oppia/no-inner-html
           temp.innerHTML = pencilCodeHtml;
+
           let output: string = '';
           let htmlObject = temp.querySelectorAll('div');
+
+          // Loop through each div element to extract its inner text.
           for (let i = 0; i < htmlObject.length; i++) {
             // eslint-disable-next-line oppia/no-inner-html
             output += htmlObject[i].innerHTML + '\n';
           }
 
+          // Mark the answer as submitted to prevent duplicate submissions.
           hasSubmittedAnswer = true;
+
+          // Submit the extracted code and its output to the current interaction service,
+          // which handles storing the response and validating it against rules.
           this.currentInteractionService.onSubmit(
             {
-              code: normalizedCode,
-              output: output || '',
-              evaluation: '',
-              error: '',
+              code: normalizedCode, // User's code
```

Comment:
We generally only use full-line comments (like those above), and not comments that come at the end of a line of code. Please fix, here and elsewhere.